src/pages/home/report/useReportUnreadMessageScrollTracking.ts

```diff
@@ -27,43 +27,94 @@ type Args = {
+        if (
+            currentVerticalScrollingOffsetRef.current > CONST.REPORT.ACTIONS.LATEST_MESSAGES_PILL_SCROLL_OFFSET_THRESHOLD &&
+            !isFloatingMessageCounterVisible &&
+            !hasUnreadMarkerReportAction
+        ) {
             setIsFloatingMessageCounterVisible(true);
         }
 
-        // hide floating button if we're scrolled closer than the offset and mark message as read
-        if (currentVerticalScrollingOffsetRef.current < CONST.REPORT.ACTIONS.SCROLL_VERTICAL_OFFSET_THRESHOLD && isFloatingMessageCounterVisible) {
-            if (readActionSkippedRef.current) {
-                // eslint-disable-next-line react-compiler/react-compiler,no-param-reassign
-                readActionSkippedRef.current = false;
-                readNewestAction(reportID);
-            }
-
+        // hide floating button if we're scrolled closer than the offset
+        if (
+            currentVerticalScrollingOffsetRef.current < CONST.REPORT.ACTIONS.LATEST_MESSAGES_PILL_SCROLL_OFFSET_THRESHOLD &&
+            isFloatingMessageCounterVisible &&
+            !hasUnreadMarkerReportAction
+        ) {
             setIsFloatingMessageCounterVisible(false);
         }
     };
 
+    const onViewableItemsChanged = useCallback(({viewableItems}: {viewableItems: ViewToken[]; changed: ViewToken[]}) => {
+        ref.current.previousViewableItems = viewableItems;
+        const viewableIndexes = viewableItems.map((viewableItem) => viewableItem.index).filter((value) => typeof value === 'number') as number[];
+        const maxIndex = Math.max(...viewableIndexes);
+        const minIndex = Math.min(...viewableIndexes);
+        const unreadActionIndex = ref.current.unreadMarkerReportActionIndex;
+        const hasUnreadMarkerReportAction = unreadActionIndex !== -1;
+        const unreadActionVisible = isInverted ? unreadActionIndex >= minIndex : unreadActionIndex <= maxIndex;
+
+        // display floating button if the unread report action is out of view
+        if (!unreadActionVisible && hasUnreadMarkerReportAction) {
+            setIsFloatingMessageCounterVisible(true);
+        }
+        // hide floating button if the unread report action becomes visible
+        if (unreadActionVisible && hasUnreadMarkerReportAction) {
+            setIsFloatingMessageCounterVisible(false);
+        }
+
+        // if we're scrolled closer than the offset and read action has been skipped then mark message as read
+        if (unreadActionVisible && readActionSkippedRef.current) {
+            // eslint-disable-next-line react-compiler/react-compiler,no-param-reassign
+            readActionSkippedRef.current = false;
+            readNewestAction(ref.current.reportID);
+        }
+
+        // FlatList requires onViewableItemsChanged callback to be stable so we can't have a dependency.
```

Comment:
```suggestion
        // FlatList requires a stable onViewableItemsChanged callback for optimal performance.
        // Therefore, we use a ref to store values instead of adding them as dependencies.

```